libsolidity/interface/StandardCompiler.cpp

```diff
@@ -50,7 +50,7 @@ namespace
 {
 
 Json::Value formatError(
-	bool _warning,
+	Error::Type _severity,
 	string const& _type,
 	string const& _component,
```

Comment:
I think the `_component` part looks like it can be inferred from severity, at least when looking down the code, it seems so.

Reply:
I don't think `component` had anything to do with `severity`?

Reply:
I'd prefer to keep that separated, might not always be possible to infer

Reply:
@cameel is it really worth it to populate `Error::Type` with cases that only happen once?

Reply:
Good question. I don't think it's a matter of frequency but more if they actually fit there. I took a deeper look at this and I no longer think they do. It seems to me that `Error` was meant specifically for problems associated with a specific location in the source code and `Error::Type` reflects that. Of the original types there the only odd one is `CodeGenerationError` for which you might not be able to pinpoint a specific location.

The extra types we have in `StandardCompiler.cpp` on the other hand are all problems that are peripheral to the compilation process and do not happen at a specific point in the code - JSON errors, unimplemented features in the codegen, assertions, mismatched checksums, errors in generated Yul code, etc. It's just that when generating the JSON response we still want to present them together with compilation errors using a unified interface, which is why `formatError()` and other functions here were not tied to `Error` and had a more flexible interface with parameters being strings. The CLI does something similar with [handling `Error` separately but in a uniform fashion](https://github.com/ethereum/solidity/blob/a532df20ec3cc7bd588381faa115a9538aebcc95/solc/CommandLineInterface.cpp#L644-L685).

So I'd say that neither these types nor something like `InterfaceError` or `GeneralError` actually belongs in `Error::Type`. I think we should backtrack on this change and keep `formatError()` only aware of the general error category, not a specific `Error::Type`. This would not require adding anything beyond `Info` to `Error::Type`. I'd suggest creating `Error::Category` enum but for now only as something you infer from `Error::Type` through a helper function - without actually splitting `Error::Type` into category and the kind of error as I suggested above.

Reply:
Yea, I actually did add the mentioned errors from `StandardCompiler` into `Error::Type`, but it does feel weird to add things like `JSONError` or `YulException` to `liblangutil`.

Yea that makes sense, but do we then rollback to having strings `JSONError`, `IOError` and etc?

Reply:
Yeah, they were already strings so I see nothing wrong with leaving them as such. We can change that later if we ever need them outside of `StandardCompiler.cpp`.